internal/cri/server/container_image_mount.go

```diff
@@ -0,0 +1,216 @@
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	containerd "github.com/containerd/containerd/v2/client"
+	"github.com/containerd/containerd/v2/core/leases"
+	"github.com/containerd/containerd/v2/core/mount"
+	"github.com/containerd/containerd/v2/defaults"
+	"github.com/containerd/errdefs"
+	"github.com/containerd/platforms"
+	"github.com/opencontainers/image-spec/identity"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+const (
+	labelGCSnapRef       = "containerd.io/gc.ref.snapshot."
+	leaseOCIVolumePrefix = "cri-oci-volume-"
+)
+
+func (c *criService) mutateMounts(
+	ctx context.Context,
+	extraMounts []*runtime.Mount,
+	sandboxID string,
+	snapshotter string,
+	platform imagespec.Platform,
+
+) error {
+	for _, m := range extraMounts {
+		err := c.mutateImageMount(ctx, m, sandboxID, snapshotter, platform)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *criService) mutateImageMount(
+	ctx context.Context,
+	extraMount *runtime.Mount,
+	sandboxID string,
+	snapshotter string,
+	platform imagespec.Platform,
+) (retErr error) {
+	imageSpec := extraMount.GetImage()
+	if imageSpec == nil {
+		return nil
+	}
+	if extraMount.GetHostPath() != "" {
+		return fmt.Errorf("hostpath must be empty while mount image: %+v", extraMount)
+	}
+	if !extraMount.GetReadonly() {
+		return fmt.Errorf("readonly must be true while mount image: %+v", extraMount)
+	}
+
+	ref := imageSpec.GetImage()
+	if ref == "" {
+		return fmt.Errorf("image not specified in: %+v", imageSpec)
+	}
+	image, err := c.LocalResolve(ref)
+	if err != nil {
+		return fmt.Errorf("failed to resolve image %q: %w", ref, err)
+	}
+	containerdImage, err := c.toContainerdImage(ctx, image)
+	if err != nil {
+		return fmt.Errorf("failed to get image from containerd %q: %w", image.ID, err)
+	}
+
+	// This is a digest of the manifest
+	imageID := containerdImage.Target().Digest.Hex()
+
+	target := c.getImageVolumeHostPath(sandboxID, imageID)
+
+	// Already mounted in another container on the same pod
+	if stat, err := os.Stat(target); err == nil && stat.IsDir() {
+		extraMount.HostPath = target
+		return nil
+	}
+
+	leaseID := leaseOCIVolumePrefix + sandboxID
+	ctx, done, err := c.client.WithLease(ctx,
+		leases.WithID(leaseID),
+		leases.WithLabel(defaults.DefaultSnapshotterNSLabel, snapshotter),
+		leases.WithLabel(labelGCSnapRef+snapshotter, target),
```

Comment:
Leases have an explicit API for holding resources, this gc reference should not be on the lease though but on the root object which owns it.

Reply:
@wzshiming I suppose since we are not "yet" supporting multiple sandboxes sharing the same sandbox image volume.. then we can gc on the destruction of the sandbox after the containers have been removed. vs creating a "shared" image with leases from the sandboxes using the image.

Reply:
I was thinking about allocating lease per sandbox. It's easy to handle the lifecycle of image volumes. We don't need to create lease on each creating request. When we delete the sandbox along with lease, all the related snapshots will be gone.